### ISSUE TITLE
Server still disconnects when transferring APK file on Android phone

### ISSUE THREAD

**Issue by efung** [claudiodangelis/qr-filetransfer#76]
I'm opening this issue because:

- [ ] I have found a bug
- [x] I want to request a feature
- [ ] I have a question
- [ ] Other

- My Go version is: `go version go1.10.3 darwin/amd64`
- My [GOPATH](https://github.com/golang/go/wiki/GOPATH) is set to: `~/go`

I'm using the latest version of the tool (I ran `go get …` last night) and saw that #13 was supposed to be resolved.

But I continue to experience problems.

I'm trying to transfer an APK file to an Android device. When I scan the code with, say, Zxing, Google Assistant/Lens, or even Twitter, I get redirected to a browser, which gives me a popup warning that the file is potentially harmful (since it's an application not downloaded from Google Play store). The connection drops at that point. Sometimes a partial download occurs. (I'm guessing the browser application downloads the file in the background, while the user answers the prompt.)

Is there anything I can do, to help you diagnose this issue?

**Comment by claudiodangelis** [comment]
Thanks for reporting, I will try to reproduce the issue and will get back
to you,

Claudio

Il mer 15 ago 2018, 19:14 Eric Fung <notifications@github.com> ha scritto:

**Comment by claudiodangelis** [comment]
I'm able to reproduce the issue and agree with what you've guessed.

As a quick workaround you can zip it before transfer, you can do that by passing the `-zip` flag:

```
qr-filtransfer -zip your.apk
```

**Comment by rafa-acioly** [comment]
the browser is canceling the download because it is a apk? i don't understand it.

**Comment by efung** [comment]
I'm not sure what's going on.

Very rarely, the transfer starts and is partially downloaded. And just
once, the entire download completed.

**Comment by efung** [comment]
> As a quick workaround you can zip it before transfer, you can do that by passing the -zip flag:

Android can't natively handle ZIP files. The end goal is to be able to install an APK directly.

**Comment by brualan** [comment]
@efung you can comment out lines `66-69` in `serveHTTP.go`. After that, the server will continue to work as long as you want, but you will need to manually shutdown it.

@claudiodangelis How about adding a flag on which the server will work as much as the user wants?

**Comment by efung** [comment]
I'll try that workaround, and like the idea of a flag to control whether the server is kept alive. It could even have a potential use-case of allowing you to transfer the same file to multiple devices.

But I'm still curious as to why the download is being aborted.

**Comment by claudiodangelis** [comment]
I'm adding the `-keep-alive` flag, sending you updates soon

**Comment by claudiodangelis** [comment]
Please check out PR #78 

```
git fetch origin
git checkout -b keep-alive origin/keep-alive
git merge master
go build
./qr-filetransfer -keep-alive=false /path/to/unknown-origin.apk
```

**Comment by efung** [comment]
Hmm. I'm not sure what's going on: I followed that script to execute `qr-filetransfer` from the PR branch. The first time, I didn't specify any `-keep-alive` flag at all, and was able to succesfully download APKs to my Android device.  I repeated using a different QR code app, and was successful again. I tried a third time, using `-keep-alive=false`, and was **still** successful. And if I specify `-keep-alive=true`, I can verify that the server is kept alive so that I can transfer again.

Not sure why the connection is being cooperative now, as opposed to last week. Perhaps I wasn't correctly installing the binary with the fix to #13 in it? Or perhaps some other change in this PR is making the transfer more reliable? `¯\_(ツ)_/¯`

**Comment by claudiodangelis** [comment]
Try renaming the file to a unique name at each trial, maybe there is some
caching going on? Weird, I'm pretty sure I've been able to reproduce the
problem you reported.

Il lun 20 ago 2018, 21:10 Eric Fung <notifications@github.com> ha scritto:

**Comment by claudiodangelis** [comment]
Any updates @efung? PR #78 apparently fixes the problem for me, but it would be great if you could confirm as well.

If everything is OK, please feel free to close this issue, thank you!

**Comment by efung** [comment]
I've tried on two different WiFi networks, and there's no consistent pattern. Sometimes, the APK is transferred fully, sometimes not.

But with the `-keep-alive=true` flag, I am always successful!

**Comment by claudiodangelis** [comment]
Sounds like an "issue fixed" to me! Please feel free to reopen the issue if it needed.

**Comment by efung** [comment]
(BTW, you can put `Resolves #76` in the PR #78 and it will auto-close this issue, when the PR is merged)

**Comment by claudiodangelis** [comment]
Oooh, I forgot about that trick. Thanks!
